optimum/onnxruntime/trainer.py

```diff
@@ -134,6 +139,30 @@
 SCALER_NAME = "scaler.pt"
 
 
+class ModuleWithLoss(nn.Module):
+    def __init__(self, model, args) -> None:
+        super().__init__()
+        self._original_model = model
+        self.args = args
+
+        # Creating an instance of huggingFace Trainer so we can use compute_loss() logic and avoid duplicated code.
+        self.hf_trainer = Trainer(model)
+        # Label smoothing
+        if self.args.label_smoothing_factor != 0:
+            from transformers.trainer_pt_utils import LabelSmoother
+
+            self.label_smoother = LabelSmoother(epsilon=self.args.label_smoothing_factor)
+        else:
+            self.label_smoother = None
```

Comment:
Why there is a case for not using label smoother? In transformers, unless using label smoother, the loss should be already calculated in the forward pass. C.f. [gpt2](https://github.com/huggingface/transformers/blob/bb5a2f2fc30985841289207b9f1f7765d8abc4e0/src/transformers/models/gpt2/modeling_gpt2.py#L1098-L1104)

Reply:
And with the wrapper, model not using label smoother in the first place shall not have any benefit on memory right?

Reply:
> And with the wrapper, model not using label smoother in the first place shall not have any benefit on memory right?

This is a good question; I can help answer it. In short, whenever label smoothing factor argument is given or not, we see improvement on memory. The key reason is the target model ORTModule wraps will have one single loss output after the change. 

If label smoothing factor argument is not given, then the CrossEntropyLoss will be done inside the model forward pass. This is true. While there is a minor tricky here: `loss` along with `lm_logits` and other intermediate states are returned in the results. If ORTModule wraps and operates on this model, during model exporting, there will be few outputs besides `loss`, those outputs in training phases are not used later, but exporter don't know it will not be used. In ORT training implementation, though those outputs are not used, but we still fill them with zero during the whole backward propagation phase.

For this case, if we wrap model+loss together, the final output of model+loss (ORTModule wraps) is just loss.

Reply:
Thanks for the explanation @pengwa, that's very clear.